### PR TITLE
chunks to lines

### PR DESCRIPTION
**Pull request by davidmarin**
In anticipation of supporting non-line-based protocols (see #715), mrjob is undergoing a transition from generators that yield lines to generators that can yield data in arbitrary chunks of bytes. When you want lines, you can wrap said chunk iterators through `mrjob.util.to_lines()`

(`to_lines()` was formerly named `buffer_iterator_to_line_iterator()`, which is more clear, but awkward to use as a wrapper function. The old name will still work until v0.6.0.)

`to_lines()` used to add a trailing newline if the last chunk didn't have one; it now leaves the data unmolested (see #819).

This also lets `mrjob.util.gunzip_stream()` just yield chunks of data rather than lines, as promised in its docstring.

I updated the implementation `to_lines()` to try to be more efficient: instead of appending to a `bytes`, we maintain a list of chunks, and `join()` them when we're ready to yield a line. Using `tests.test_util.ReadFileTestCase.test_large_bz2_file()` as a benchmark, I determined that... the old and new implementations are just about equally fast. I still believe it's probably a bit better for the cases I care about, but can't prove it yet.

### CODE REVIEW

File: mrjob/util.py [hunk 1]
```diff
@@ -739,3 +684,44 @@ def unarchive(archive_path, dest):
...
+            else:
+                yield chunk[start:end]
+
+            start = end
+
+    if leftovers:
+        yield b''.join(leftovers)
+
+#: Deprecated alias for :py:func:`yield_lines`, will be removed in v0.6.0
+buffer_iterator_to_line_iterator = yield_lines
```
Comment (jcnnghm) [inline mrjob/util.py]: Should calls to `buffer_iterator_to_line_iterator` warn?

File: tests/test_util.py [hunk 2]
```diff
@@ -48,34 +49,49 @@ class BufferIteratorToLineIteratorTestCase(TestCase):
...
+            list(yield_lines(_ for _ in ())),
             [])
 
     def test_buffered_lines(self):
         self.assertEqual(
-            list(buffer_iterator_to_line_iterator(chunk for chunk in
-                                                  [b'The quick\nbrown fox\nju',
-                                                   b'mped over\nthe lazy\ndog',
-                                                   b's.\n'])),
+            list(yield_lines(chunk for chunk in
```
Comment (jcnnghm) [inline tests/test_util.py]: I think it's more important to maintain a consistent style, but we've been formatting new code like:

```
self.assertEqual(
    list(
        yield_lines(
            chunk for chunk in
            [
                b'The quick\nbrown fox\nju',
                b'mped over\nthe lazy\ndog', 
                b's.\n'
            ]
        )
    ),
    [b'The quick\n', b'brown fox\n', b'jumped over\n', b'the lazy\n', b'dogs.\n']
)
```
Feel free to keep formatting as-is, I don't feel strongly about it either way.

**Comment by jcnnghm** [comment]
Code looks good to me, no blockers.  Feel free to merge.

File: mrjob/util.py (hunk 1 above)
Comment (davidmarin) [inline mrjob/util.py]: Good point. Reworked this so that a warning is issued.

File: tests/test_util.py (hunk 2 above)
Comment (davidmarin) [inline tests/test_util.py]: Cool. mrjob uses PEP8 for style (see the [release checklist](https://github.com/Yelp/mrjob/wiki/Release-Checklist).

Does Yelp have a maximum line length?

**Comment by davidmarin** [comment]
Okay, cool. After sleeping on it, decided `to_lines()` was a better (mostly, shorter) name than `yield_lines()`.

Other than making `buffer_iterator_to_line_iterator()` warn, the other changes are just renames and documentation, so planning to merge if tests pass.

**Comment by jcnnghm** [comment]
Sounds good to me.

**Pull request merged**